src/entity/vertex.hpp

```diff
@@ -0,0 +1,43 @@
+#include "../types.hpp"
+
+#include <glm/glm.hpp>
+
+#include <functional>
+
+namespace entity {
+
+struct Vertex {
+    VoxelOffset position;
+    glm::i8vec3 normal;
+    MatColorId mat_color_id;
+    uint8_t ambient_occlusion;
+
+    bool
+    operator==(const Vertex& other) const {
+        return (
+            other.position == position && other.normal == normal
+            && other.mat_color_id == mat_color_id
+            && other.ambient_occlusion == ambient_occlusion
+        );
+    }
+};
+
+} // namespace entity
+
+template <>
+struct std::hash<entity::Vertex> {
+    size_t
+    operator()(const entity::Vertex& vertex) const {
+        size_t position_hash = static_cast<uint32_t>(
+            vertex.position.x ^ vertex.position.y ^ vertex.position.z
+        );
+
+        /* clang-format off */
+        uint8_t normal_hash = (vertex.normal.x << 2)
+                            | (vertex.normal.y << 1)
+                            | (vertex.normal.z);
+
+        return position_hash ^ vertex.mat_color_id ^ normal_hash ^ vertex.ambient_occlusion;
+        /* clang-format on */
+    }
```

Comment:
Use the Boost `hash_combine` function

```cpp
template <class T>
inline void hash_combine(std::size_t& seed, const T& v)
{
    std::hash<T> hasher;
    seed ^= hasher(v) + 0x9e3779b9 + (seed<<6) + (seed>>2);
}
```

Like so:
```suggestion
        size_t result = 0;

        // Position
        utils::hash_combine(result, vertex.position.x);
        utils::hash_combine(result, vertex.position.y);
        utils::hash_combine(result, vertex.position.z);

        // Normal
        utils::hash_combine(result, vertex.normal.x << 2)
        utils::hash_combine(result, vertex.normal.y << 1)
        utils::hash_combine(result, vertex.normal.z)

        // Color ID and ambient occlusion
        utils::hash_combine(result, vertex.mat_color_id);
        utils::hash_combine(result, vertex.ambient_occlusion);

        return result
    }
```